rosidl_generator_py/resource/_msg_support.c.template

```diff
@@ -0,0 +1,160 @@
+#include <Python.h>
+
+#include <@(spec.base_type.pkg_name)/@(subfolder)/@(module_name)__struct.h>
+#include <@(spec.base_type.pkg_name)/@(subfolder)/@(module_name)__functions.h>
+
+@{
+have_not_included_primitive_arrays = True
+have_not_included_string = True
+}@
+@[for field in spec.fields]@
+@[if field.type.is_array and have_not_included_primitive_arrays]@
+@{have_not_included_primitive_arrays = False}@
+#include <rosidl_generator_c/primitives_array.h>
+#include <rosidl_generator_c/primitives_array_functions.h>
+
+@[end if]@
+@[if field.type.type == 'string' and have_not_included_string]@
+@{have_not_included_string = False}@
+#include <rosidl_generator_c/string.h>
+#include <rosidl_generator_c/string_functions.h>
+
+@[end if]@
+@[end for]@
+
+@{
+msg_typename = '%s__%s__%s' % (spec.base_type.pkg_name, subfolder, spec.base_type.type)
+}@
+
+/* TODO(esteve): implement conversion function */
+void * @(spec.base_type.pkg_name)_@(module_name)__convert_from_py(PyObject *_pymsg)
+{
+  @(msg_typename) * ros_message = @(msg_typename)__create();
+  (void)ros_message;
+@{
+full_classname = "@(spec.base_type.pkg_name).@(subfolder)._@(module_name).@(spec.base_type.type)"
+}@
+  char full_classname_dest[@(len(full_classname) + 1)];
+
+  /* TODO(esteve): assume 1BYTE encoding for unicode strings. */
+  char * class_name = (char *)PyUnicode_1BYTE_DATA(
+    PyObject_GetAttrString(PyObject_GetAttrString(_pymsg, "__class__"), "__name__"));
+  char * module_name = (char *)PyUnicode_1BYTE_DATA(
+    PyObject_GetAttrString(PyObject_GetAttrString(_pymsg, "__class__"), "__module__"));
+
+  snprintf(full_classname_dest, @(len(full_classname) + 1), "%s.%s", class_name, module_name);
+
+  assert(strncmp(
+    "@(spec.base_type.pkg_name).@(subfolder)._@(module_name).@(spec.base_type.type)",
+    full_classname_dest, @(len(full_classname))));
```

Comment:
I assume this is trying to check that the type name matches. But `strncmp` will return `0` if the strings are equal and `assert` wants `true`?

Since the check only looks at `len` (and not at `len +1`) it would also fail if `full_classname_dest` is just a prefix of the constructed string.